### ISSUE TITLE
Nx 16.8.0+ no longer automatically adds the 'apps' or 'libs' prefix for app/lib generators

### ISSUE THREAD

**Issue by simondotm** [simondotm/nx-firebase#180]
See https://nx.dev/deprecated/as-provided-vs-derived

For Nx 17, there are migration options for derived names, but they say from Nx 18 it will be as-provided only.

Apparently we have to provide `--directory` for workspace structure ourselves now.

Anyone finding this a problem with their projects? 

I'm testing the plugin at the moment, and e2e tests are failing but I think the plugin should work ok with this new Nx change, just need to see if the documentation for the plugin may need updating to reflect this.

**Comment by simondotm** [comment]
In the old way, you could create projects in a sub directory and Nx would name that project accordingly.
So `nx g @nx/js:lib my-proj --directory=subdir`
Would make a project in `libs/subdir/my-proj` and call it `subdir-my-proj` (but project filename would be `my-proj.json`) which was handy.

Now, we cant do this, we have to do:
`nx g @nx/js:lib my-proj --directory=apps/subdir`
and the project is called `my-proj` (not `subdir-my-proj`)

If we try this:
`nx g @nx/js:lib subdir-my-proj --directory=apps/subdir`
then the project is called `subdir-my-proj` (but filename of project is also `subdir-my-proj.json`)

🤔

**Comment by thekip** [comment]
Update: it seems NX trying to be smart and do some heuristic over the monorepo. So if you have "app" folder it will create new applications in here. So my previous report valid only if you already have apps / libs folder created.

**Comment by simondotm** [comment]
Ok I'm going to close this. My plugin compatibility tests seem ok with Nx versions >=16.8.1
The next plugin release will require a minimum 16.8.1 Nx version, so can be addressed if anyone reports a related issue.

**Comment by simondotm** [comment]
Re-opening this.
Just trying to update plugin to Nx 16.8.1 and having problems with my e2e tests. I don't think the plugin functions & app generators are respecting the new options for generators:
* `--projectNameAndRootFormat` should be parsed & used if provided
* `--directory` should consider if the `nx.json` file specifies `as-provided` or `derived` and derive the destination based on name/directory accordingly

Hopefully theres some tooling for this in `@nx/devkit`

**Comment by simondotm** [comment]
> Update: it seems NX trying to be smart and do some heuristic over the monorepo. So if you have "app" folder it will create new applications in here. So my previous report valid only if you already have apps / libs folder created.

@thekip I believe there's some fangled new configs in `nx.json` that might be influencing this:

```json
  "workspaceLayout": {
    "projectNameAndRootFormat": "derived",
    "appsDir": "apps",
    "libsDir": "libs"
  },
``` 
🤔 
Nx 16.8 blog post [here](https://blog.nrwl.io/nx-16-8-release-e38e3bb503b5#b353)

**Comment by simondotm** [comment]
The default `nx.json` created by the e2e test framework is just:
```json
  "workspaceLayout": {
    "projectNameAndRootFormat": "as-provided"
  },
```
I'll probably need to add some e2e tests to make sure we've got functional coverage for all the different workspace variants.

**Comment by simondotm** [comment]
This is now implemented in #205
